core/include/drivers/pl061_gpio.h

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright (c) 2016, Linaro Limited
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __PL061_GPIO_H__
+#define __PL061_GPIO_H__
+
+#include <gpio.h>
+
+void pl061_gpio_register(vaddr_t base_addr, int gpio_dev);
```

Comment:
Assuming that by 'id' you mean the `gpio_dev` arg?

The id actually simplifies the rest of the code a lot. We can use a simplified way, i.e. just use `gpio / GPIOS_PER_PL061` to calculate the index into array of registered base addresses to find the right base address to use. It does put the burden on the platform (code) to register/index the gpio devices properly, but it should be easy to do, i.e. the first device is 0, 2nd device is 1, 3rd dev 2, and so on so forth.

Without the id, and by just using `base address`, we would still need some sort of indexed array anyway, usually within a `struct` somewhere internally within the pl061 driver code, to keep track of all the available devices, and repeated loopings through the array to calculate the proper base address required for every get_ and set_ calls based on the gpio pin/number provided.

This was actually how it was done originally [1][2] until ARM proposed the simplification [3].

[1] https://github.com/96boards-hikey/arm-trusted-firmware/blob/hikey/drivers/arm/gpio/gpio.c#L118-L132
[2] https://github.com/ARM-software/arm-trusted-firmware/pull/510/commits/fe8880cc0771539fb7112f838bd93fb312af2be7#diff-ee806812453e4998cf59d65d6ceab1afR150
[3] https://github.com/hzhuang1/arm-trusted-firmware/commit/b688de6c94485847b839f6225eb36c128e66d3a1#diff-ee806812453e4998cf59d65d6ceab1afR137